File: src/main/java/dev/kikugie/techutils/feature/worldedit/WorldEditNetworkHandler.java
package dev.kikugie.techutils.feature.worldedit;

import dev.kikugie.techutils.TechUtilsMod;
import net.fabricmc.fabric.api.client.networking.v1.ClientPlayNetworking;
import net.minecraft.client.MinecraftClient;
import net.minecraft.network.packet.CustomPayload;
import net.minecraft.util.Identifier;
import net.minecraft.util.math.BlockPos;
import org.enginehub.worldeditcui.protocol.CUIPacket;

import java.util.Optional;

/**
 * Opens a WorldEdit CUI channel and passes packets to {@link WorldEditSync} instance.
 * <p>
 * If channel has already been registered by another mod, it leeches packets from it instead.
 */
public class WorldEditNetworkHandler {
	public static final Identifier CHANNEL = Identifier.of("worldedit", "cui");
	private static final int PROTOCOL = 4;
	private static WorldEditNetworkHandler instance;
	public final WorldEditStorage storage;
	private boolean worldEditConnected = false;
	private boolean yoinkPackets = false;

	private WorldEditNetworkHandler() {
		this.storage = WorldEditStorage.init();
		var receivers = ClientPlayNetworking.getGlobalReceivers();
		if (receivers.contains(CHANNEL)) {
			this.yoinkPackets = true;
			handshake();
			return;
		}
		ClientPlayNetworking.registerReceiver(CUIPacket.TYPE, (payload, context) -> this.onPacket(payload));
		handshake();
	}

	public static WorldEditNetworkHandler initHandler() {
		if (MinecraftClient.getInstance().world == null) {
			throw new RuntimeException("Registering WorldEdit handler in non-world context!");
		}
		instance = new WorldEditNetworkHandler();
		return instance;
	}

	public static Optional<WorldEditNetworkHandler> getInstance() {
		return Optional.ofNullable(instance);
	}

	public void onYoinkedPacket(CustomPayload payload) {
		if (!this.yoinkPackets) return;
		onPacket((CUIPacket) payload);
	}

	private void onPacket(CUIPacket packet) {
		if (!this.worldEditConnected) {
			this.worldEditConnected = true;
			TechUtilsMod.LOGGER.info("WorldEdit connected");
			WorldEditSync.getInstance().ifPresent(WorldEditSync::onWorldEditConnected);
		}
		if (packet.args().isEmpty()) {
			TechUtilsMod.LOGGER.warn("WorldEditNetworkHandler#onPacket(): Received CUI packet of length zero");
			return;
		}
		handlePacket(packet.eventType(), packet.args().toArray(String[]::new));
	}

	private void handlePacket(String type, String[] args) {
		if (type.equals("s")) {
			this.storage.setCuboid(args[0].equals("cuboid"));
			return;
		}
		if (!type.equals("p") || !this.storage.isCuboid())
			return;
		try {
			int p = Integer.parseInt(args[0]);
			BlockPos pos = new BlockPos(Integer.parseInt(args[1]), Integer.parseInt(args[2]), Integer.parseInt(args[3]));
			this.storage.setPos(p == 0, pos);
		} catch (NumberFormatException e) {
			TechUtilsMod.LOGGER.warn("WorldEditCUINetworkHandler#handlePacket(): Failed int parsing of position", e);
		}
	}

	private void handshake() {
		ClientPlayNetworking.send(new CUIPacket("v", String.valueOf(PROTOCOL)));
	}
}

File: src/client/kotlin/dev/kikugie/techutils/impl/worldedit/WorldEditNetworkHandler.kt
package dev.kikugie.techutils.impl.worldedit

import com.mojang.brigadier.CommandDispatcher
import dev.kikugie.techutils.TechUtilsClient
import io.netty.buffer.Unpooled
import net.fabricmc.fabric.api.client.networking.v1.ClientPlayNetworking
import net.minecraft.client.MinecraftClient
import net.minecraft.command.CommandSource
import net.minecraft.network.PacketByteBuf
import net.minecraft.util.Identifier
import net.minecraft.util.math.BlockPos
import java.nio.charset.StandardCharsets


class WorldEditNetworkHandler {
    val storage = WorldEditStorage()
    var connected = false
    var available = false
    var yoinkPackets = false
    var initialized = false

    val shouldTick
        get() = connected && available

    init {
        requireNotNull(MinecraftClient.getInstance().world) {
            "Registering WorldEdit handler in non-world context"
        }
        if (CHANNEL in ClientPlayNetworking.getGlobalReceivers())
            yoinkPackets = true
        else {
            ClientPlayNetworking.registerReceiver(CHANNEL) { _, _, it, _ -> onDataPacket(it) }
            handshake()
        }
    }

    fun tick() {
        if (initialized) return
        if (shouldTick) WorldEditSync.onConnected()
        initialized = shouldTick
    }

    fun onCommandTreePacket(dispatcher: CommandDispatcher<CommandSource>) {
        available = dispatcher.findNode(listOf("/pos1")) != null
    }

    fun onDataPacket(data: PacketByteBuf) {
        if (!connected) {
            TechUtilsClient.LOGGER.info("WorldEdit connected!")
            connected = true
        }

        val bytes = data.readableBytes()
        if (bytes == 0) {
            TechUtilsClient.LOGGER.warn("Received CUI packet of length zero")
            return
        }
        val message = data.toString(0, data.readableBytes(), StandardCharsets.UTF_8)
        val split = message.split('|')
        val multi = split[0].startsWith("+")
        val type = split[0].substring(if (multi) 1 else 0)
        val args = message.substring(type.length + (if (multi) 2 else 1)).split('|').toTypedArray()
        TechUtilsClient.LOGGER.debug(message)
        handlePacket(type, *args)
    }

    private fun handlePacket(type: String, vararg args: String) {
        if (args.isEmpty())
            TechUtilsClient.LOGGER.info("Received CUI packet with no arguments")
        else if (type == "s")
            storage.cuboid = args.first() == "cuboid"
        else if (type == "p" && storage.cuboid) try {
            val index = args[0].toInt()
            val pos = BlockPos(args[1].toInt(), args[2].toInt(), args[3].toInt())
            if (index == 0) storage.pos1 = pos
            else storage.pos2 = pos
        } catch (e: Exception) {
            TechUtilsClient.LOGGER.warn("Failed to parse CUI args [${args.joinToString(", ")}], type $type", e)
        }
    }

    private fun handshake() {
        val message = "v|$PROTOCOL"
        val buf = Unpooled.copiedBuffer(message, StandardCharsets.UTF_8)
        ClientPlayNetworking.send(CHANNEL, PacketByteBuf(buf))
    }

    companion object {
        val CHANNEL = Identifier("worldedit", "cui")
        val PROTOCOL = 4

        var INSTANCE: WorldEditNetworkHandler? = null
            private set

        fun init() {
            INSTANCE = WorldEditNetworkHandler()
        }

        fun close() {
            INSTANCE = null
        }
    }
}